Tests/Adyen Tests/Analytics/AnalyticsProviderTests.swift

```diff
@@ -163,6 +192,129 @@ class AnalyticsProviderTests: XCTestCase {
+        
+        let analyticsResponse = AnalyticsResponse()
+        let analyticsResult: Result<Response, Error> = .success(analyticsResponse)
+        
+        apiClient.mockedResults = [checkoutAttemptIdResult, analyticsResult]
+        
+        let expectation = expectation(description: "should not be called")
+        expectation.isInverted = true
+        
+        apiClient.onExecute = { _ in
+            expectation.fulfill()
+        }
+        
+        sut.sendInitialAnalytics(with: .components(type: .achDirectDebit), additionalFields: nil)
+        sut.sendEventsIfNeeded()
+        
+        wait(for: [expectation], timeout: 0.1)
+    }
+    
+    func testEventShouldSendWhenAttemptIdAndEventsExist() {
+        let apiClient = APIClientMock()
+        let sut = AnalyticsProvider(apiClient: apiClient,
+                                    configuration: AnalyticsConfiguration(),
+                                    eventDataSource: eventDataSource)
+
+        let initialAnalyticsResponse = InitialAnalyticsResponse(checkoutAttemptId: checkoutAttemptIdMockValue)
+        let checkoutAttemptIdResult: Result<Response, Error> = .success(initialAnalyticsResponse)
+        
+        let analyticsResponse = AnalyticsResponse()
+        let analyticsResult: Result<Response, Error> = .success(analyticsResponse)
+        
+        apiClient.mockedResults = [checkoutAttemptIdResult, analyticsResult]
+        
+        sut.sendInitialAnalytics(with: .components(type: .achDirectDebit), additionalFields: nil)
+        wait(for: .milliseconds(100))
+        
+        sut.add(info: AnalyticsEventInfo(component: "card", type: .rendered))
+        sut.sendEventsIfNeeded()
+        wait(for: .milliseconds(100))
+        
+        // events return nil if empty
+        XCTAssertNil(sut.eventDataSource.wrappedEvents())
+    }
+    
+    func testAddingLogEventShouldTriggerSend() {
+        let apiClient = APIClientMock()
+        let sut = AnalyticsProvider(apiClient: apiClient,
+                                    configuration: AnalyticsConfiguration(),
+                                    eventDataSource: eventDataSource)
+
+        let initialAnalyticsResponse = InitialAnalyticsResponse(checkoutAttemptId: checkoutAttemptIdMockValue)
+        let checkoutAttemptIdResult: Result<Response, Error> = .success(initialAnalyticsResponse)
+        
+        let analyticsResponse = AnalyticsResponse()
+        let analyticsResult: Result<Response, Error> = .success(analyticsResponse)
+        
+        apiClient.mockedResults = [checkoutAttemptIdResult, analyticsResult]
+        
+        sut.sendInitialAnalytics(with: .components(type: .achDirectDebit), additionalFields: nil)
+        wait(for: .milliseconds(100))
+        
+        sut.add(log: AnalyticsEventLog(component: "card", type: .submit, subType: .sdk))
+        wait(for: .milliseconds(100))
+        
+        // events return nil if empty
+        XCTAssertNil(sut.eventDataSource.wrappedEvents())
```

Comment:
I would doubt this test actually tests analytics provider behavior. Assuming datasource is faulty and just fails to store any events this test seem to be green! This test checks for rather datasource side-effects and not the sut behavior. Testing side-effects might be fine but here I see we can improve this.
I suggest testing the `apiClient.perform` and check the request data if possible.

And the second this that we test here is "datasource is cleaned up". This could be tested by injecting a datasource mock rather than using a real class and then checking if `removeAllEvents` was called.